core/src/main/java/org/htsjdk/core/utils/PathSpecifier.java

```diff
@@ -0,0 +1,238 @@
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public boolean isPath() {
+        try {
+            return getCachedPath() != null || toPath() != null;
+        } catch (ProviderNotFoundException |
+                FileSystemNotFoundException |
+                IllegalArgumentException |
+                HtsjdkException |
+                AssertionError e) {
+            // jimfs throws an AssertionError that wraps a URISyntaxException when trying to create path where
+            // the scheme-specific part is missing or incorrect
+            pathFailureReason = e.getMessage();
+            return false;
+        }
+    }
+
+    @Override
+    public URI getURI() {
+        return uri;
+    }
+
+    @Override
+    public String getURIString() {
+        return getURI().toString();
+    }
+
+    /**
+     * Return the raw input string as provided to the constructor.
+     */
+    @Override
+    public String getRawInputString() { return rawInputString; }
+
+    /**
+     * Resolve the URI to a {@link Path} object.
+     *
+     * @return the resulting {@code Path}
+     * @throws HtsjdkException if an I/O error occurs when creating the file system
+     */
+    @Override
+    public Path toPath() {
+        if (getCachedPath() != null) {
+            return getCachedPath();
+        } else {
+            final Path tmpPath = Paths.get(getURI());
+            setCachedPath(tmpPath);
+            return tmpPath;
+        }
+    }
+
+    @Override
+    public String getToPathFailureReason() {
+        if (pathFailureReason == null) {
+            try {
+                toPath();
+                return String.format("'%s' appears to be a valid Path", rawInputString);
+            } catch (ProviderNotFoundException e) {
+                return String.format("ProviderNotFoundException: %s", e.getMessage());
+            } catch (FileSystemNotFoundException e) {
+                return String.format("FileSystemNotFoundException: %s", e.getMessage());
+            } catch (IllegalArgumentException e) {
+                return String.format("IllegalArgumentException: %s", e.getMessage());
+            } catch (HtsjdkException e) {
+                return String.format("UserException: %s", e.getMessage());
+            }
+        }
+        return pathFailureReason;
+    }
+
+    @Override
+    public InputStream getInputStream() {
+        if (!isPath()) {
+            throw new HtsjdkException(getToPathFailureReason());
+        }
+
+        final Path resourcePath = toPath();
+        try {
+            return Files.newInputStream(resourcePath);
+        } catch (IOException e) {
+            throw new HtsjdkException(
+                    String.format("Could not create open input stream for %s (as URI %s)", getRawInputString(), getURIString()), e);
+        }
+    }
+
+    @Override
+    public OutputStream getOutputStream() {
+        if (!isPath()) {
+            throw new HtsjdkException(getToPathFailureReason());
```

Comment:
No raw `HtsjdkException`

Reply:
Yeah - perhaps we need a separate discussion about how we want to partition the exception namespace. https://github.com/samtools/htsjdk-next-beta/issues/36.